File: tests/test_pipeline.py
import threading
import time
from pathlib import Path

from simple_pipe.context.frame_meta import FrameInput
from simple_pipe.pipeline.simple_pipeline import SimplePipeline
from simple_pipe.spec.runtime_config import PipelineRuntimeConfig

ROOT = Path(__file__).resolve().parents[1]


def _run_frames(pipeline: SimplePipeline, count: int = 2, timeout: float = 5.0):
    results = []
    done = threading.Event()

    def on_output(meta):
        results.append(meta)
        if len(results) >= count:
            done.set()

    pipeline.set_output_callback(on_output)
    assert pipeline.start().ok
    for i in range(count):
        assert pipeline.push_frame(FrameInput(image="x", frame_index=i)).ok
    assert done.wait(timeout), "timed out waiting for sink callback"
    pipeline.stop()
    return results


def test_linear_pipeline():
    pipeline = SimplePipeline(defaults_dir=ROOT / "configs/demo/node_defaults")
    assert pipeline.build_from_json(ROOT / "configs/demo/graph/linear.json").ok
    results = _run_frames(pipeline, count=2)
    assert results[0].extensions.get("resize", {}).get("width") == 320


def test_safety_fanin_pipeline():
    pipeline = SimplePipeline()
    assert pipeline.build_from_json(ROOT / "configs/demo/graph/safety_fanin.json").ok
    pipeline.configure(PipelineRuntimeConfig.from_dict({"node_params": {"rules": {"min_score": 0.85}}}))
    results = _run_frames(pipeline, count=2)
    assert "safety" in results[-1].extensions
    assert "tracks" in results[-1].extensions

File: tests/test_pipeline.cpp
#include <atomic>
#include <chrono>
#include <thread>

#include <gtest/gtest.h>

#include "simple_pipe/context/frame_meta.hpp"
#include "simple_pipe/pipeline/simple_pipeline.hpp"
#include "simple_pipe/spec/runtime_config.hpp"
#include "test_helper.hpp"

using namespace simple_pipe;
using namespace simple_pipe::test;

namespace {

int RunFrames(SimplePipeline& pipeline, int count, int timeout_ms = 5000) {
  std::atomic<int> received{0};
  pipeline.SetOutputCallback([&](const FrameMeta&) { received.fetch_add(1); });
  if (!pipeline.StartPipeline().IsSuccess()) {
    return -1;
  }

  for (int i = 0; i < count; ++i) {
    FrameInput input;
    input.image = "frame";
    input.frame_index = i;
    if (!pipeline.PushFrame(input).IsSuccess()) {
      pipeline.StopPipeline();
      return -1;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(2));
  }

  const auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeout_ms);
  while (received.load() < count && std::chrono::steady_clock::now() < deadline) {
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  pipeline.StopPipeline();
  return received.load();
}

}  // namespace

TEST(PipelineTest, LinearGraphProducesOutput) {
  SimplePipeline pipeline(RepoPath("configs/demo/node_defaults").string());
  ASSERT_TRUE(pipeline.BuildFromJson(GraphPath("linear")).IsSuccess());

  nlohmann::json runtime = {
      {"node_params", {{"resize", {{"width", 320}, {"height", 240}}}}},
  };
  ASSERT_TRUE(pipeline.Configure(PipelineRuntimeConfig::FromJson(runtime)).IsSuccess());
  EXPECT_EQ(RunFrames(pipeline, 2), 2);
}

TEST(PipelineTest, SafetyFanInGraphProducesSafetyExtension) {
  SimplePipeline pipeline;
  ASSERT_TRUE(pipeline.BuildFromJson(GraphPath("safety_fanin")).IsSuccess());

  std::atomic<bool> saw_safety{false};
  pipeline.SetOutputCallback([&](const FrameMeta& meta) {
    if (meta.extensions.contains("safety")) {
      saw_safety.store(true);
    }
  });
  ASSERT_TRUE(pipeline.StartPipeline().IsSuccess());
  FrameInput input;
  input.image = "x";
  input.frame_index = 0;
  ASSERT_TRUE(pipeline.PushFrame(input).IsSuccess());

  std::this_thread::sleep_for(std::chrono::milliseconds(200));
  pipeline.StopPipeline();
  EXPECT_TRUE(saw_safety.load());
}
